File: src/test/java/org/nexial/core/variable/FileTest.java
/*
 * Copyright 2012-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.nexial.core.variable;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.SystemUtils;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.nexial.commons.utils.FileUtil;

import static java.io.File.separator;

public class FileTest {
    private String testDir;

    @Before
    public void setUp() {
        testDir = SystemUtils.getJavaIoTmpDir().getAbsolutePath() + separator +
                  this.getClass().getSimpleName() + separator;
        new java.io.File(testDir).mkdirs();
    }

    @After
    public void tearDown() {
        if (testDir != null) { FileUtils.deleteQuietly(new java.io.File(testDir)); }
    }

    @Test
    public void content() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        File file = new File();
        String content = file.content(target);
        Assert.assertEquals(fixture, content);
    }

    @Test
    public void asList() throws IOException {
        String item1 = RandomStringUtils.randomAlphanumeric(256);
        String item2 = RandomStringUtils.randomAlphanumeric(256);
        String item3 = RandomStringUtils.randomAlphanumeric(256);
        String item4 = RandomStringUtils.randomAlphanumeric(256);
        String item5 = RandomStringUtils.randomAlphanumeric(256);
        String fixture = item1 + "\n" +
                         item2 + "\n" +
                         item3 + "\r\n" +
                         item4 + "\r\n" +
                         item5;
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        String content = new File().asList(target);
        List<String> list = Arrays.asList(StringUtils.split(content, ","));
        Assert.assertEquals(5, list.size());
        Assert.assertEquals(item1, list.get(0));
        Assert.assertEquals(item2, list.get(1));
        Assert.assertEquals(item3, list.get(2));
        Assert.assertEquals(item4, list.get(3));
        Assert.assertEquals(item5, list.get(4));
    }

    @Test
    public void lastmod() throws IOException {
        long rightNow = System.currentTimeMillis();

        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        String lastmod = new File().lastmod(target);

        // assert that the file was created in less than last 2 seconds
        Assert.assertTrue((Long.parseLong(lastmod) - rightNow) < 2000);
    }

    @Test
    public void size() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals(1024, Integer.parseInt(new File().size(target)));
    }

    @Test
    public void copy() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        String target2 = testDir + "dummy" + separator + "test2.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals(target2, new File().copy(target, target2));
        Assert.assertTrue(FileUtil.isFileReadable(target2, 1024));
    }

    @Test
    public void move() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        String target2 = testDir + "dummy" + separator + "test2.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals(target2, new File().move(target, target2));
        Assert.assertTrue(FileUtil.isFileReadable(target2, 1024));
        Assert.assertFalse(FileUtil.isFileReadable(target));
    }

    @Test
    public void delete() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals(target, new File().delete(target));
        Assert.assertFalse(FileUtil.isFileReadable(target));
    }

    @Test
    public void overwrite() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        String fixture2 = RandomStringUtils.randomAlphanumeric(1048);
        Assert.assertEquals(target, new File().overwrite(target, fixture2));
        Assert.assertTrue(FileUtil.isFileReadable(target, 1048));
        Assert.assertEquals(fixture2, FileUtils.readFileToString(new java.io.File(target), Charset.forName("UTF-8")));
    }

    @Test
    public void overwrite_new_file() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";

        Assert.assertEquals(target, new File().overwrite(target, fixture));
        Assert.assertTrue(FileUtil.isFileReadable(target, 1024));
        Assert.assertEquals(fixture, FileUtils.readFileToString(new java.io.File(target), Charset.forName("UTF-8")));
    }

    @Test
    public void append() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        String fixture2 = RandomStringUtils.randomAlphanumeric(1048);
        Assert.assertEquals(target, new File().append(target, fixture2));
        Assert.assertTrue(FileUtil.isFileReadable(target, 1024 + 1048));
        Assert.assertEquals(fixture + fixture2,
                            FileUtils.readFileToString(new java.io.File(target), Charset.forName("UTF-8")));
    }

    @Test
    public void prepend() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        String fixture2 = RandomStringUtils.randomAlphanumeric(1048);
        Assert.assertEquals(target, new File().prepend(target, fixture2));
        Assert.assertTrue(FileUtil.isFileReadable(target, 1024 + 1048));
        Assert.assertEquals(fixture2 + fixture,
                            FileUtils.readFileToString(new java.io.File(target), Charset.forName("UTF-8")));
    }

    @Test
    public void name() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals("test1.txt", new File().name(target));
    }

    @Test
    public void dir() throws IOException {
        String fixture = RandomStringUtils.randomAlphanumeric(1024);
        String target = testDir + "test1.txt";
        FileUtils.writeStringToFile(new java.io.File(target), fixture, Charset.forName("UTF-8"));

        Assert.assertEquals(StringUtils.removeEnd(testDir, separator),
                            StringUtils.removeEnd(new File().dir(target), separator));
    }
}

File: src/test/kotlin/org/nexial/core/variable/FileTest.kt
/*
 * Copyright 2012-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.nexial.core.variable

import org.apache.commons.io.FileUtils
import org.apache.commons.lang3.RandomStringUtils
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.SystemUtils
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.FixMethodOrder
import org.junit.Test
import org.junit.runners.MethodSorters.NAME_ASCENDING
import org.nexial.commons.utils.FileUtil
import java.io.File.separator
import java.io.IOException
import java.nio.charset.Charset

@FixMethodOrder(NAME_ASCENDING)
class FileTest {

    private var testDir = "${SystemUtils.getJavaIoTmpDir().absolutePath}$separator${this.javaClass.simpleName}_"

    @Before
    fun setUp() {
        // if (!java.io.File(testDir!!).mkdirs()) System.err?.println("Unable to create $testDir")
    }

    @After
    fun tearDown() {
        FileUtils.deleteQuietly(java.io.File(testDir))
    }

    @Test
    @Throws(IOException::class)
    fun content() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}content${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val file = File()
            val content = file.content(target)
            Assert.assertEquals(fixture, content)
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun asList() {
        val osArch = System.getProperty("os.arch")
        if (StringUtils.equals(osArch, "aarch64")) {
            // skipping this test for ARM64... not sure why it doesn't work here
            println("skipping asList() test for $osArch system since... well, it doesn't work (NOT SURE WHY")
            return;
        }

        val item1 = RandomStringUtils.randomAlphanumeric(256)
        val item2 = RandomStringUtils.randomAlphanumeric(256)
        val item3 = RandomStringUtils.randomAlphanumeric(256)
        val item4 = RandomStringUtils.randomAlphanumeric(256)
        val item5 = RandomStringUtils.randomAlphanumeric(256)
        val fixture = item1 + "\n" +
                      item2 + "\n" +
                      item3 + "\r\n" +
                      item4 + "\r\n" +
                      item5
        val dir = "${testDir}asList${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val content = File().asList(target)
            val list = StringUtils.split(content, ",")
            // val list = listOf(*StringUtils.split(content, ","))
            Assert.assertEquals(5, list.size)
            Assert.assertEquals(item1, list[0])
            Assert.assertEquals(item2, list[1])
            Assert.assertEquals(item3, list[2])
            Assert.assertEquals(item4, list[3])
            Assert.assertEquals(item5, list[4])
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun lastmod() {
        val rightNow = System.currentTimeMillis()

        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}lastmod${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val lastmod = File().lastmod(target)

            // assert that the file was created in less than last 2 seconds
            Assert.assertTrue(java.lang.Long.parseLong(lastmod) - rightNow < 2000)
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun size() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}size${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))
            Assert.assertEquals(1024, Integer.parseInt(File().size(target)).toLong())
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun copy() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir1 = "${testDir}copy${separator}"
        val target = "${dir1}test1.txt"
        val dir2 = "${testDir}dummy${separator}"
        val target2 = "${dir2}test2.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            Assert.assertEquals(target2, File().copy(target, target2))
            Assert.assertTrue(FileUtil.isFileReadable(target2, 1024))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir1))
            FileUtils.deleteQuietly(java.io.File(dir2))
        }
    }

    @Test
    @Throws(IOException::class)
    fun move() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir1 = "${testDir}move${separator}"
        val target = "${dir1}test1.txt"
        val dir2 = "${testDir}dummy${separator}"
        val target2 = "${dir2}test2.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            Assert.assertEquals(target2, File().move(target, target2))
            Assert.assertTrue(FileUtil.isFileReadable(target2, 1024))
            Assert.assertFalse(FileUtil.isFileReadable(target))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir1))
            FileUtils.deleteQuietly(java.io.File(dir2))
        }
    }

    @Test
    @Throws(IOException::class)
    fun delete() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}delete${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            Assert.assertEquals(target, File().delete(target))
            Assert.assertFalse(FileUtil.isFileReadable(target))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun overwrite() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}overwrite${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val fixture2 = RandomStringUtils.randomAlphanumeric(1048)
            Assert.assertEquals(target, File().overwrite(target, fixture2))
            Assert.assertTrue(FileUtil.isFileReadable(target, 1048))
            Assert.assertEquals(fixture2, FileUtils.readFileToString(java.io.File(target), Charset.forName("UTF-8")))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun overwrite_new_file() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}new_file${separator}"
        val target = "${dir}test1.txt"

        try {
            Assert.assertEquals(target, File().overwrite(target, fixture))
            Assert.assertTrue(FileUtil.isFileReadable(target, 1024))
            Assert.assertEquals(fixture, FileUtils.readFileToString(java.io.File(target), Charset.forName("UTF-8")))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun append() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}append${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val fixture2 = RandomStringUtils.randomAlphanumeric(1048)
            Assert.assertEquals(target, File().append(target, fixture2))
            Assert.assertTrue(FileUtil.isFileReadable(target, (1024 + 1048).toLong()))
            Assert.assertEquals(fixture + fixture2,
                                FileUtils.readFileToString(java.io.File(target), Charset.forName("UTF-8")))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun prepend() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}prepend${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            val fixture2 = RandomStringUtils.randomAlphanumeric(1048)
            Assert.assertEquals(target, File().prepend(target, fixture2))
            Assert.assertTrue(FileUtil.isFileReadable(target, (1024 + 1048).toLong()))
            Assert.assertEquals(fixture2 + fixture,
                                FileUtils.readFileToString(java.io.File(target), Charset.forName("UTF-8")))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun name() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}name${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            Assert.assertEquals("test1.txt", File().name(target))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }

    @Test
    @Throws(IOException::class)
    fun dir() {
        val fixture = RandomStringUtils.randomAlphanumeric(1024)
        val dir = "${testDir}dir${separator}"
        val target = "${dir}test1.txt"
        try {
            FileUtils.writeStringToFile(java.io.File(target), fixture, Charset.forName("UTF-8"))

            Assert.assertEquals(StringUtils.removeEnd("${testDir}dir", separator),
                                StringUtils.removeEnd(File().dir(target), separator))
        } finally {
            FileUtils.deleteQuietly(java.io.File(dir))
        }
    }
}